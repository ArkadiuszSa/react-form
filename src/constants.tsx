class Constants {
  public API_BASE;
  constructor() {
    this.API_BASE = "https://event-list-server.herokuapp.com/api";
  }
}

export default new Constants();
