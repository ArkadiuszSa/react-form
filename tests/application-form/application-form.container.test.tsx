import * as React from "react";
import ApplicationFormContainer from "../../src/modules/user/pages/application-form/application-form.container";
import { shallow, mount } from "enzyme";
import configureMockStore from "redux-mock-store";
import * as enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as moment from "moment";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import constants from "../../src/constants";
const mockStore = configureMockStore();
enzyme.configure({ adapter: new Adapter() });

const apiBase = constants.API_BASE;

let mock = new MockAdapter(axios);

describe("ApplicationForm-container, integration tests betwen container,component and actions", () => {
  const INITIAL_STATE = {
    applicationForm: {
      title: "",
      firstName: "",
      firstNameErr: "",
      lastName: "",
      lastNameErr: "",
      email: "",
      emailErr: "",
      date: "",
      dateErr: "",
      submitInfo: "",
      avaibleDates: [],
      selectedDate: null
    }
  };
  let wrapper, store;

  beforeEach(() => {
    store = mockStore(INITIAL_STATE);
    wrapper = shallow(
      <ApplicationFormContainer
        store={store}
        match={{ params: { id: "id" } }}
      />,
      {
        lifecycleExperimental: true
      }
    );
  });

  test("maps state and dispatch to props", () => {
    expect(wrapper.props().title).toMatchSnapshot();
    expect(wrapper.props().firstName).toMatchSnapshot();
    expect(wrapper.props().firstNameErr).toMatchSnapshot();
    expect(wrapper.props().lastName).toMatchSnapshot();
    expect(wrapper.props().lastNameErr).toMatchSnapshot();
    expect(wrapper.props().email).toMatchSnapshot();
    expect(wrapper.props().emailErr).toMatchSnapshot();
    expect(wrapper.props().date).toMatchSnapshot();
    expect(wrapper.props().dateErr).toMatchSnapshot();
    expect(wrapper.props().submitInfo).toMatchSnapshot();
    expect(wrapper.props().avaibleDates).toMatchSnapshot();
    expect(wrapper.props().selectedDate).toMatchSnapshot();
  });

  test("should run RESET_APPLICATION_FORM action on componentWillMount()", () => {
    wrapper.dive();
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test("should run FETCH_APPLICATION_HAPPENING with payload", done => {
    mock.onGet(apiBase + "/happening/id").reply(200, {
      title: "Happening Name",
      days: ["2018-11-11"]
    });

    wrapper.props().fetchHappening("id");
    wrapper.update();

    setTimeout(() => {
      const actions = store.getActions();
      expect(actions).toMatchSnapshot();
      done();
    }, 0);
  });

  test("should run  RESET_APPLICATION_FORM action and  SUBMIT_APPLICATION_FORM_FAILURE action form with errors payload ", () => {
    const fakeEvent = { preventDefault: () => {} };

    wrapper
      .dive()
      .find("form")
      .simulate("submit", fakeEvent);

    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test("should also run HANDLE_FORM_CHANGE action with payload and RESET_APPLICATION_FORM_SUBMIT_INFO ", () => {
    wrapper = mount(
      <ApplicationFormContainer
        store={store}
        match={{ params: { id: "id" } }}
      />
    );

    wrapper
      .find("Input")
      .at(0)
      .props()
      .onChange({ target: { name: "firstName", value: "John" } });

    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test("should also run HANDLE_DATE_CHANGE action with payload", () => {
    wrapper.props().handleDateChange(moment("2018-11-11"));
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test("should submit valid form and use SUBMIT_APPLICATION_FORM_SUCCES action with payload ", () => {
    mock.onPost(apiBase + "/application").reply(200, {});
    wrapper.props().submitApplicationForm(
      { preventDefault: () => {} },
      {
        firstName: "John",
        lastname: "Doo",
        email: "johndoo@gmail.com",
        date: "2018-11-11",
        avaibleDates: [moment("2018-11-11")]
      }
    );

    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test("should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to user was signed in with that email ", done => {
    mock.onPost(apiBase + "/application").reply(400, {
      errors: ["This email is already signed up for this event"]
    });

    wrapper.props().submitApplicationForm(
      { preventDefault: () => {} },
      {
        firstName: "John",
        lastname: "Doo",
        email: "johndoo@gmail.com",
        date: "2018-11-11",
        avaibleDates: [moment("2018-11-11")]
      }
    );

    setTimeout(() => {
      const actions = store.getActions();
      expect(actions).toMatchSnapshot();
      done();
    }, 0);
  });

  test("should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to emial validation on api ", done => {
    mock.onPost(apiBase + "/application").reply(400, {
      errors: [
        {
          value: "error value",
          param: "email"
        }
      ]
    });

    wrapper.props().submitApplicationForm(
      { preventDefault: () => {} },
      {
        firstName: "John",
        lastname: "Doo",
        email: "johndoo@gmail.com",
        date: "2018-11-11",
        avaibleDates: [moment("2018-11-11")]
      }
    );

    setTimeout(() => {
      const actions = store.getActions();
      expect(actions).toMatchSnapshot();
      done();
    }, 0);
  });

  test("should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to other error ", done => {
    mock.onPost(apiBase + "/application").reply(404, {
      errors: ["other error"]
    });

    wrapper.props().submitApplicationForm(
      { preventDefault: () => {} },
      {
        firstName: "John",
        lastname: "Doo",
        email: "johndoo@gmail.com",
        date: "2018-11-11",
        avaibleDates: [moment("2018-11-11")]
      }
    );

    setTimeout(() => {
      const actions = store.getActions();
      expect(actions).toMatchSnapshot();
      done();
    }, 0);
  });
});
