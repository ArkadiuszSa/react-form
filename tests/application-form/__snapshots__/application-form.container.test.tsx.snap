// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 1`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 2`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 3`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 4`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 5`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 6`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 7`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 8`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 9`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 10`] = `""`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 11`] = `Array []`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions maps state and dispatch to props 12`] = `null`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should also run HANDLE_DATE_CHANGE action with payload 1`] = `
Array [
  Object {
    "payload": Object {
      "date": "2018-11-11",
      "selectedDate": "2018-11-10T23:00:00.000Z",
    },
    "type": "HANDLE_DATE_CHANGE",
  },
  Object {
    "type": "RESET_APPLICATION_FORM_SUBMIT_INFO",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should also run HANDLE_FORM_CHANGE action with payload and RESET_APPLICATION_FORM_SUBMIT_INFO  1`] = `
Array [
  Object {
    "type": "RESET_APPLICATION_FORM",
  },
  Object {
    "payload": Object {
      "name": "firstName",
      "value": "John",
    },
    "type": "HANDLE_FORM_CHANGE",
  },
  Object {
    "type": "RESET_APPLICATION_FORM_SUBMIT_INFO",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should run  RESET_APPLICATION_FORM action and  SUBMIT_APPLICATION_FORM_FAILURE action form with errors payload  1`] = `
Array [
  Object {
    "type": "RESET_APPLICATION_FORM",
  },
  Object {
    "payload": Object {
      "dateErr": "Requires date",
      "emailErr": "Requires valid email",
      "firstNameErr": "Requires first name",
      "lastNameErr": "Requires last name",
    },
    "type": "VALIDATION_APPLICATION_FORM_FAILURE",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should run FETCH_APPLICATION_HAPPENING with payload 1`] = `
Array [
  Object {
    "payload": Object {
      "avaibleDates": Array [
        "2018-11-10T23:00:00.000Z",
      ],
      "date": "2018-11-11",
      "selectedDate": "2018-11-10T23:00:00.000Z",
      "title": "Sign up for Happening Name",
    },
    "type": "FETCH_APPLICATION_HAPPENING_SUCCES",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should run RESET_APPLICATION_FORM action on componentWillMount() 1`] = `
Array [
  Object {
    "type": "RESET_APPLICATION_FORM",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to emial validation on api  1`] = `
Array [
  Object {
    "type": "VALIDATION_APPLICATION_FORM_SUCCES",
  },
  Object {
    "payload": "The provided email is invalid!",
    "type": "SUBMIT_APPLICATION_FORM_FAILURE",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to other error  1`] = `
Array [
  Object {
    "type": "VALIDATION_APPLICATION_FORM_SUCCES",
  },
  Object {
    "payload": "Form has not been saved correctly! Please check the connection.",
    "type": "SUBMIT_APPLICATION_FORM_FAILURE",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should submit form, pass validation and  and use SUBMIT_APPLICATION_FORM_FAILURE action with payload due to user was signed in with that email  1`] = `
Array [
  Object {
    "type": "VALIDATION_APPLICATION_FORM_SUCCES",
  },
  Object {
    "payload": "This email is already signed up for this event!",
    "type": "SUBMIT_APPLICATION_FORM_FAILURE",
  },
]
`;

exports[`ApplicationForm-container, integration tests betwen container,component and actions should submit valid form and use SUBMIT_APPLICATION_FORM_SUCCES action with payload  1`] = `
Array [
  Object {
    "type": "VALIDATION_APPLICATION_FORM_SUCCES",
  },
]
`;
